### PR TITLE
feat(snapshots): implementation of snapshot garbage collector

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**

### CODE REVIEW

**Comment by Abhinandan-Purkait** [comment]
bors try

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+impl GarbageCollect for OperationGuardArc<VolumeSnapshot> {
+    async fn garbage_collect(&mut self, context: &PollContext) -> PollResult {
+        self.destroy_deleting(context).await
+    }
+
+    async fn destroy_deleting(&mut self, context: &PollContext) -> PollResult {
+        deleting_volume_snapshot_reconciler(self, context).await
+    }
+
+    // Unimplemented garbage collectors
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs:65]: Could we not implement some of these?

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#[tracing::instrument(skip(snapshot, context), level = "trace", fields(snapshot.id = %snapshot.uuid(), request.reconcile = true))]
+async fn deleting_volume_snapshot_reconciler(
+    snapshot: &mut OperationGuardArc<VolumeSnapshot>,
+    context: &PollContext,
+) -> PollResult {
+    if !snapshot.as_ref().status().deleting() {
+        return Ok(PollerState::Idle);
+    }
+
+    let snap_user_spec = snapshot.lock().spec().clone();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: don't need to clone here, only within the if let Some?

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    if let Some(snap_rsc) = context.specs().volume_snapshot_rsc(snapshot.uuid()) {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: can replace with let Some(snap_rsc) else { } to remove nesting

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        match snapshot
+            .destroy(
+                context.registry(),
+                &DestroyVolumeSnapshotRequest::new(snap_rsc, snap_user_spec),
+            )
+            .await
+        {
+            Ok(_) => {
+                tracing::info!("VolumeSnapshot {:?} deleted successfully", snapshot.uuid());
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: pass uuid on the left side

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
+use crate::{
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs:1]: Add a garbage collector reconciler as TaskPoller and set poll period to 5

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 6]
```diff
@@ -629,6 +629,12 @@ impl ResourceSpecsLocked {
         specs.volume_snapshots.get(snapshot_id).cloned()
     }
 
+    /// Gets a copy of all locked VolumeSnapshots.
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs]:
```suggestion
    /// Gets a copy of all resourced VolumeSnapshots.
```

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs (hunk 3 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: Where do the `if block` go in this? Sorry I am not aware of this syntax.

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs (hunk 2 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: Right

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs (hunk 3 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: no if block, like so..
```rust
let Some(snap_rsc) = context.specs().volume_snapshot_rsc(snapshot.uuid()) else {
  return Ok(PollerState::Idle);
}
```

Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: Done.

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs (hunk 5 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs:1]: Done.

**Review by abhilashshetty04 — APPROVED** [review]

**Review by tiagolobocastro — APPROVED** [review]

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**